gossip/api/types.go

```diff
@@ -0,0 +1,229 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package api describes the external interface to the gossip hub.
+package api
+
+import (
+	"crypto/sha256"
+	"fmt"
+
+	"github.com/google/certificate-transparency-go/tls"
+)
+
+// PathPrefix is the common prefix for all HTTP entrypoints.
+const PathPrefix = "/gossip/v0/"
+
+// Add Log Head to Hub
+// POST https://<hub server>/gossip/v0/add-log-head
+// Inputs:
+//    source_url: The URL from which the log head was obtained
+//    head_data: The binary-encoded head
+//    log_signature: The source log's signature over head_data.
+// Outputs:
+//    timestamped_entry: TLS-encoded TimestampedEntry.
+//    hub_signature: hub's signature gossip_timestamped_entry
+
+// AddLogHeadPath is the final path component for this entrypoint.
+const AddLogHeadPath = "add-log-head"
+
+// AddLogHeadRequest represents the JSON request body sent to the add-log-head POST method.
+type AddLogHeadRequest struct {
+	SourceURL    string `json:"source_url"`
+	HeadData     []byte `json:"head_data"`
+	LogSignature []byte `json:"log_signature"`
+}
+
+// AddLogHeadResponse represents the JSON response body for the add-log-head POST method.
+type AddLogHeadResponse struct {
+	// TimestampedEntryData holds the TLS-encoding of a TimestampedEntry.
+	TimestampedEntryData []byte `json:"timestamped_entry"`
+	// HubSignature is a signature over GossipTimestamp.
+	HubSignature []byte `json:"hub_signature"`
+}
+
+// TimestampedEntry holds a timestamped entry in the hub, in a form suitable for TLS-encoding.
+// RFC 5246 notation:
+//   struct {
+//      opaque source_url<1,65535>;
```

Comment:
Not sure we should include the source_url here since a blob can be made available over multiple URLs. 
It does sound like it would be useful as auxillary data though

Reply:
Is `source_url` stable enough to be used as a long term ID for the sources of signed blobs?
Perhaps we should include something more directly related to the public key

`source_url` useful regardless since it may help investigators replicate results.

Reply:
Shifted to `SourceID` to be more generic.